python/tank/templatekey.py

```diff
@@ -292,31 +304,164 @@ def __init__(self,
+
+        # If there is a format and strict_matching is set, there's an error, since there
+        # is no format to enforce or not.
+        if self.format_spec is None and strict_matching is not None:
+            raise TankError("strict_matching can't be set if there is no format_spec")
+
+        # By default, if strict_matching is not set but there is a format spec, we'll
+        # strictly match.
+        if strict_matching is None and self.format_spec is not None:
+            strict_matching = True
+
+        if strict_matching:
+            # This regular expression is blind to the actual length of the string for performance
+            # reasons. Code that uses it should test that the string's length is of
+            # self._minimum_width first. It first matches up to n-1 0s. It then matches either a
+            # single 0, or an actual two digit number that doesn't start with 0.
+            self._strict_validation_re = re.compile("^%s{0,%d}((%s)|0)$" % (
+                self._padding_char, self._minimum_width - 1, self._NON_ZERO_POSITIVE_INTEGER_EXP)
+            )
+        else:
+            self._strict_validation_re = None
+
+        self.strict_matching = strict_matching
+
     def validate(self, value):
+        """
+        Validates the value.
+
+        :param value: Value to validate.
 
+        :returns: True is the validation was succesfull, False otherwise.
+        """
         if value is not None:
-            if not (isinstance(value, int) or value.isdigit()):
-                self._last_error = "%s Illegal value %s, expected an Integer" % (self, value)
+            if isinstance(value, basestring):
+                # We have a string, make sure it loosely or strictly matches the format.
+                if self.strict_matching and not self._strictly_matches(value):
+                    return False
+                elif not self.strict_matching and not self._loosely_matches(value):
+                    return False
+            elif not isinstance(value, int):
+                self._last_error = "%s Illegal value '%s', expected an Integer" % (self, value)
                 return False
-            else:
-                return super(IntegerKey, self).validate(value)
+            return super(IntegerKey, self).validate(value)
+        return True
+
+    def _loosely_matches(self, value):
+        """
+        Checks if the value loosely matches. The value loosely matches if it can be turned into an
+        int.
+
+        :param value: String to test
+
+        :returns: True if it loosely matches, False otherwise.
+        """
+        # This is the extent of what was tested before strict matching. We're actually a bit more permissive,
+        # because the user could have specified a format specifier that uses spaces for padding, but isdigit will
+        # fail if spaces are at the beginning of a string, so strip them out.
+        if self._padding_char == ' ':
```

Comment:
this may not be right, but containing the riff i started above in the code, maybe this could be 
`if not self._zero_padded:`? So tracking the *state* of the format spec (zero padded true vs false) rather than the logic of *how* the format spec should be processed (e.g. padding char '0' vs padding char ' ')? Curious to hear your thoughts!